sky/execution.py

```diff
@@ -261,6 +261,9 @@ def _execute(
         # NOTE: in general we may not have sufficiently specified info
         # (cloud/resource) to check STOP_SPOT_INSTANCE here. This is checked in
         # the backend.
+        if any(r.ports is not None for r in task.resources):
```

Comment:
I don't remember if `task.resources` being a set means `any_of` is being used. (We should add a comment to Task.resources.) In that case, we can have `any_of { ..requesting ports.., ..not requesting ports.. }`, and the logic here will effectively disallow the second resource.

Is the above right? Wdyt is the best way to handle this?

Reply:
Actually it won't disallow the second resources.  For the following YAML:

```yaml
# a.yaml
resources:
  any_of:
    - ports: 8080
      accelerators: A100-80GB:1
    - accelerators: A100-80GB:2
```

`sky launch` will select and only select the second resources, e.g. A100-80GB:2:

```bash
$ SKYPILOT_DEBUG=0 sky launch a.yaml
Task from YAML spec: a.yaml
I 02-13 15:35:09 optimizer.py:1215] No resource satisfying <Cloud>({'A100-80GB': 1}, ports=['8080']) on RunPod.
I 02-13 15:35:09 optimizer.py:691] == Optimizer ==
I 02-13 15:35:09 optimizer.py:702] Target: minimizing cost
I 02-13 15:35:09 optimizer.py:714] Estimated cost: $4.0 / hour
I 02-13 15:35:09 optimizer.py:714] 
I 02-13 15:35:09 optimizer.py:837] Considered resources (1 node):
I 02-13 15:35:09 optimizer.py:907] ---------------------------------------------------------------------------------------------------
I 02-13 15:35:09 optimizer.py:907]  CLOUD    INSTANCE              vCPUs   Mem(GB)   ACCELERATORS   REGION/ZONE   COST ($)   CHOSEN   
I 02-13 15:35:09 optimizer.py:907] ---------------------------------------------------------------------------------------------------
I 02-13 15:35:09 optimizer.py:907]  RunPod   2x_A100-80GB_SECURE   16      160       A100-80GB:2    CA            3.98          ✔     
I 02-13 15:35:09 optimizer.py:907] ---------------------------------------------------------------------------------------------------
I 02-13 15:35:09 optimizer.py:907] 
Launching a new cluster 'sky-67bc-root'. Proceed? [Y/n]: ^CAborted!
```

Reply:
The above experiment is run on a new docker container and only enabled runpod

Reply:
Yeah you are right... And it seems like we don't need the handle here since we will add resources-specific cloud required features and remove any related cloud for each resources in current master: https://github.com/skypilot-org/skypilot/blob/2f0432bff3a70d264f4b2cfde91220c9670c98fa/sky/clouds/cloud.py#L341-L353

Reply:
Actually here we should only add task-level features like stop and multi-node